### PR TITLE
Dynamic multiapp

### PR DESCRIPTION
**Pull request by permcody**
Support for Dynamic Multiapps!

Here's a good first cut at this new feature. Comments are welcome.

A few things that need to be completed before this can be merged:
- [x] Fix issue with dependency_libs not be populated in the modules when compiled separately as noted in #4643.
- [x] Test and see what include guards, TestHarness caveats, extra error messages are needed when compiled in static mode
- [x] Linux tests, different METHOD tests

Tasks that need to be completed after this is merged:
- Add new registration function to existing applications
- Stork update for new registration method

### CODE REVIEW

**Comment by friedmud** [comment]
Awesome!  I'm going to take some time to look through this... tomorrow (absolutely blasted today...)

**Comment by permcody** [comment]
I ran into a nasty issue when compiling this in static mode. I received several duplicate symbol errors for "registerApps()" which makes sense. Each of the libraries needed for the final executable will contain that global unmangled symbol when I use `extern "C"`. Rather than make any attempt to conditionally export that extern declaration, I think the most portable path forward will be to instead make those function names unique. I plan to use this form

```extern "C" void AnimalApp__registerApps() { AnimalApp::registerApps(); } ```
(i.e. I plan to replace the colons in the scoped C++ name with underscores in the C-compatible name)

**Comment by andrsd** [comment]
If there is a preprocessor macro that would tell you about dynamic/static build, you can do something like:
```
#ifdef STATIC_BUILD
#define EXPORT static
#else
#define EXPORT
#endif
```
Then you use it:
```
EXPORT extern "C" void registerApps()
```
The static keyword would hide this symbol in static builds, thus preventing duplicate symbols in linking. Since we should not need to call this in a static build, all would be fine. Right?

**Comment by permcody** [comment]
AFAIK - there isn't any standard compiler flag since static/dynamic mode is
technically a linker flag. That being said, we could rig up something in
libMesh but it would be libMesh specific.

**Comment by andrsd** [comment]
The build system knows if it is doing static or dynamic build and then it would use `-DSTATIC_BUILD` which is standard compiler flag. That's how I saw it being done. I do not know if libMesh build system has such a macro or not. Just saying this is another solution to the problem...

**Comment by friedmud** [comment]
I don't understand - registerApps() should be in MOOSE... so it shouldn't
actually be in the application libraries at all.  So it shouldn't be an
issue

Admittedly: I still haven't had time to look at this closely yet...

Derek

**Comment by permcody** [comment]
registerApps() is in every single application/module it has to be so you
have something to call when you load it.

**Comment by permcody** [comment]
I thought for a moment that this might be a header/body issue. I verified that it is in fact, _not_ the issue.

**Comment by permcody** [comment]
This now fails gracefully when compiled in static mode.

**Comment by permcody** [comment]
@idaholab/moose-developers - I just updated the test to make sure that we really are picking up a different module (i.e. PhaseFieldApp) as the subapp in this test case. This PR runs on OS X and Linux in both dynamic and static modes. Runs fine with the various libMesh METHODS. It'll fail gracefully with a static compile saying it can't fine a library containing the requested App. Take a look at it then push the big green button if all y'all are happy with it.

**Comment by friedmud** [comment]
I still want to take a look at this.

Sorry I'm gumming up the works here...

File: framework/src/multiapps/MultiApp.C [hunk 1]
```diff
@@ -512,3 +528,140 @@ MultiApp::globalAppToLocal(unsigned int global_app)
...
+  pcrecpp::RE re_deps("(/\\S*\\.la)");
+
+  std::ifstream handle(library_filename.c_str());
+  if (handle.is_open())
+  {
+    while (std::getline(handle, line))
+    {
+      // Look for the system dependent dynamic library filename to open
+      if (line.find("dlname=") != std::string::npos)
+        // Magic numbers are computed from length of this string "dlname=' and line minus that string plus quotes"
```
Comment (friedmud) [inline framework/src/multiapps/MultiApp.C]: Lol - ok.  At least there is a comment :-)

**Comment by friedmud** [comment]
Pretty sweet overall.  A couple of things:

1.  Relying on a capitilization scheme is definitely a little brittle.  Should we add options to override all of thr mangling you're doing in case someone has something named weirdly?

2.  There should be no need for having `AppName__registerApps` have the `AppName` stuff on the front of it.  There are lots of options here including just putting this over in a separate file that doesn't get compiled into the static library at all.  Something like `MoosePlugin.h` Or something.  The reason I'm thinking about it is that I would like to be able to provide lots of types of plugins (Kernels, BCs, etc.) and not just Apps... so we would need a few of these functions... bot just Apps.  Doing it with manual mangling is "ok"... but a little ugly.


Overall this is pretty awesome though!

**Comment by permcody** [comment]
1. I was already thinking about providing a library name override in case the name mangling doesn't work. I'll probably beef up the error message when it can't find the multiapp too like printing out the paths searched and the filename it was attempting to look for. This will save on support issues.

2. So I did think about trying to split out the registration function however libMesh actually supports building both static and dynamic libraries at the same time using the same object files. It would take even more Makefile magic to make that work that I'm not in the mood to fight with. We could just "not" support that case but then users who compile MOOSE with just "--enable-static" would get a nasty link message - yuck. Besides, I'm not 100% convinced we wouldn't run into some other odd problem with one of our compiler chains with these duplicate global symbols all being linked into a single executable. If all you're concerned about is having additional functions, I would propose changing the signature of the `Appname__registerApps()` method instead. We could for instance go with this: `Appname__register(SYS_TYPE type)`. That would give us an entry point to dispatch to any number of functions to accomplish anything.

**Comment by permcody** [comment]
Both comments addressed.

**Comment by friedmud** [comment]
I actually think I prefer the old registration myself.  If you want to stick with pseudo mangling (which is fine with me - I was just throwing out ideas).  Then I would go with multiple functions that match up with what's in MooseApp:

`AppName__registerApps()`
`AppName__registerObjects()`
`AppName__associateSyntax()`

The latter two would allow you to "pull in" another application's objects and syntax (or even just register a single object that is sitting in a tiny little plugin library) by passing in _this_ object's Factory, Syntax and ActionFactory.

**Comment by permcody** [comment]
That's fine too...  I put this in an extra commit just in case we don't like it. I'll hold off on adding the other stubs for now in case we decide to add more (like register single object) as we make these new capabilities concrete. I can peel off the last commit and then have it merged unless there are any other opinions.

**Comment by friedmud** [comment]
I think we should add the `registerObjects` and `associateSyntax` now.  One thing we can do with it is take the explicit registration of Modules out of applications.

Instead: we could just have:

```
[Modules]
  use = 'PhaseField SolidMechanics HeatConduction'
[]
```

And the system would just pick those up and call their registration functions automatically... the sky is really the limit here...

Anyway - we don't have to think that far ahead: but I do think it would be good to get the three stubs in there from the beginning...

**Comment by friedmud** [comment]
Oh, one more thing: this kind of busts Peacock a bit.  Is there anything we can do so that Peacock can get the set of available MultiApps?

Maybe something like search through `MOOSE_LIBRARY_PATH` and load up each one and try to call it's registration function and then return that out as the possible Apps for MultiApps?

**Comment by permcody** [comment]
I suggest merging soon to give it a trial run. If everything goes smoothly
then we can move towards that. The big issue is that people may forget to
compile and post issues on the list. "Why is my app not working after I run
make clean and rebuild?"

**Comment by friedmud** [comment]
Anyway - that's probably beyond the scope of this PR.

Let's get this merged then put in more tickets...

**Comment by friedmud** [comment]
Heh - agreed.  Can you peel off that last commit?  Then I'll hit merge :-)

**Comment by permcody** [comment]
At, lunch. But will do shortly

**Pull request merged**